### PR TITLE
[WIP] Introduce particle property pool

### PR DESCRIPTION
**Pull request by gassmoeller**
Rebase of #1220. Still lacking the changes discussed on Monday with @bangerth.

### CODE REVIEW

**Comment by bangerth** [comment]
I had a look at this earlier today, and I think you actually have everything in place. In particular, you have the `translation_table`. All you have to do when you free an entry instead (i) move the contents of the last slot entry into that slot, (ii) free the last slot instead (can be implemented cheaply using `IndexSet`), (iii) redirect the entry in the `translation_table` that points to the last used slot to the slot whose memory was supposed to be freed.

The last step may be expensive if you store this as a vector as you do right now. Whether it matters is a different question -- I suspect it probably won't, but in case it does: What you need is a *bimap* that allows you to search for both keys and values with O(log N) effort. I haven't looked but suspect that boost may have one.

**Comment by bangerth** [comment]
This also doesn't build on the tester because the tester uses deal.II 8.4.1.

We may have to hold off on this until after the next release. Or, we could at least settle on the interfaces by having the implementation ready, but if we use an old deal.II using a trivial pool that simply uses `new`/`delete` as we currently do by hand.

**Comment by bangerth** [comment]
Let me work on the pre-8.5 version of the property pool -- I need to do some programming every once in a while...

**Comment by gassmoeller** [comment]
Feel free to go ahead. 
There are still some issues with the current state, I was trying to fix (but do not know if I still manage that today, I also still need to pack my stuff for tomorrow). Problems arise in particular when the particles carry no properties. In this case `PropertyPool` returns an `ArrayView(Null,0)`, but the ArrayView copy constructor throws an Assert when used in this situation, so `const ArrayView properties = particle.get_properties();` is not allowed. We need to figure out what to do in that case. I feel like it should be allowed to ask a particle without properties for its properties (it should simply return an empty ArrayView or something).

**Comment by bangerth** [comment]
Yes, we should relax `ArrayView`. Since we need newest deal.II anyway, there is no real reason not to change `ArrayView` -- a view of size 0 to a non-existent array should definitely make sense.

**Comment by bangerth** [comment]
I'm curious to see whether this now works!

File: source/particle/property_pool.cc [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    {}
+
+
+
+    PropertyPool::Handle
+    PropertyPool::allocate_properties_array ()
+    {
+#if DEAL_II_VERSION_GTE(8,5,0)
+      if (n_properties == 0)
+        return numbers::invalid_unsigned_int;
```
Comment (bangerth) [inline source/particle/property_pool.cc]: return `invalid_handle` here.

File: source/particle/property_pool.cc [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+          translation_table.resize(size);
+
+          IndexSet old_used_slots(free_slots.size());
+          old_used_slots.add_range(0,free_slots.size());
+          old_used_slots.subtract_set(free_slots);
+
+          free_slots.clear();
+          free_slots.set_size(size);
+          free_slots.add_range(0,size);
+          free_slots.subtract_set(old_used_slots);
```
Comment (bangerth) [inline source/particle/property_pool.cc]: simplify this if the old slots were all used.

File: include/aspect/particle/property_pool.h [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+      public:
+#if DEAL_II_VERSION_GTE(8,5,0)
+        typedef IndexSet::size_type Handle;
+#else
+        typedef double *Handle;
+#endif
+        static const Handle invalid_handle;
+
+
+        PropertyPool (const unsigned int n_properties_per_slot);
```
Comment (bangerth) [inline include/aspect/particle/property_pool.h]: add documentation to all of these members

File: include/aspect/particle/property_pool.h [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        void reserve(const std::size_t size);
+
+        unsigned int n_properties_per_slot () const;
+
+      private:
+        unsigned int n_properties;
+
+#if DEAL_II_VERSION_GTE(8,5,0)
+        std::vector<double> memory_pool;    // size == n_slots * n_properties_per_slot;
+        std::vector<std::vector<double>::size_type> translation_table; // size = n_slots, content: Addresses within memory_pool
```
Comment (bangerth) [inline include/aspect/particle/property_pool.h]: use a map here

**Comment by bangerth** [comment]
@gassmoeller -- want to rebase this? Or close it and open a github issue to remind us of the unfinished business?

It also looks like this doesn't compile right now because the version of boost that was bundled with 8.4 didn't include `bimap.hpp` :-(

**Comment by gassmoeller** [comment]
This is no longer relevant as the file and classes are now in deal.II

**Pull request closed without merging**